### PR TITLE
lib/ramfs: Add comments to ramfs.h file

### PR DESCRIPTION
**Pull request by DeliaPavel**
Signed-off-by: Delia-Maria Pavel <delia_maria.pavel@stud.acs.upb.ro>

### Prerequisite checklist

 - [ ] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [ ] Tested your changes against relevant architectures and platforms;
 - [ ] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]

### CODE REVIEW

File: lib/ramfs/ramfs.h [hunk 1]
```diff
@@ -48,17 +48,40 @@ struct ramfs_node {
...
 
+/**
+ * Allocates ramfs node
+ *
+ * @param name
+ *   the name of the file/directory
+ * @param type
+ *   the type (file or directory)
+ * @return
+ *   the ramfs node
```
Comment (StefanJum) [inline lib/ramfs/ramfs.h]:
```suggestion
 *   the new ramfs_node structure
```
I would refer to the structure with it's exact name (`ramfs_node`) so Doxygen will link it to the `ramfs_node` structure page.

File: lib/ramfs/ramfs.h [hunk 2]
```diff
@@ -48,17 +48,40 @@ struct ramfs_node {
 	size_t rn_size;    /* file size */
 	char *rn_buf;    /* buffer to the file data */
 	size_t rn_bufsize;    /* allocated buffer size */
-	struct timespec rn_ctime;
-	struct timespec rn_atime;
-	struct timespec rn_mtime;
-	int rn_mode;
-	bool rn_owns_buf;
+	struct timespec rn_ctime;  /* last change time */
```
Comment (StefanJum) [inline lib/ramfs/ramfs.h]: Please add the structure member comments in a comment block above the actual member declaration.
```c
/**
 * last change time
 */
struct timespec rn_ctime;
```

File: lib/ramfs/ramfs.h [hunk 3]
```diff
@@ -48,17 +48,40 @@ struct ramfs_node {
...
+	struct timespec rn_atime; /* last access time */
+	struct timespec rn_mtime; /* last modification time */
+	int rn_mode; /* node access mode */
+	bool rn_owns_buf; /* if the rn_buf was allocated in this ramfs_node */
 };
 
+/**
+ * Allocates ramfs node
```
Comment (StefanJum) [inline lib/ramfs/ramfs.h]:
```suggestion
 * Allocates ramfs node.
```

File: lib/ramfs/ramfs.h [hunk 4]
```diff
@@ -48,17 +48,40 @@ struct ramfs_node {
...
+ *
+ * @param name
+ *   the name of the file/directory
```
Comment (StefanJum) [inline lib/ramfs/ramfs.h]:
```suggestion
 *   The name of the file/directory
```
Please start the description with a capital letter. Same goes for all the other param/return comments

File: lib/ramfs/ramfs.h [hunk 5]
```diff
@@ -48,17 +48,40 @@ struct ramfs_node {
...
+ * @param type
+ *   the type (file or directory)
+ * @return
+ *   the ramfs node
+ */
 struct ramfs_node *ramfs_allocate_node(const char *name, int type);
 
+/**
+ * Frees ramfs node
```
Comment (StefanJum) [inline lib/ramfs/ramfs.h]:
```suggestion
 * Frees a ramfs node.
```

File: lib/ramfs/ramfs.h [hunk 6]
```diff
@@ -48,17 +48,40 @@ struct ramfs_node {
...
+ *
+ * @param node
+ *   the node that will be freed
+ */
 void ramfs_free_node(struct ramfs_node *node);
 
+/**
+ * Transforms a vnode into a ramfs_node
```
Comment (StefanJum) [inline lib/ramfs/ramfs.h]:
```suggestion
 * Transforms a vnode into a ramfs_node.
```

**Review by StefanJum — CHANGES_REQUESTED** [review]
Thank you @DeliaPavel, just some minor comments.

File: lib/ramfs/ramfs.h [hunk 7]
```diff
@@ -36,29 +36,65 @@
 #include <vfscore/prex.h>
 #include <stdbool.h>
 
-/*
- * File/directory node for RAMFS
+/**
```
Comment (razvand) [inline lib/ramfs/ramfs.h:39]: Do not use dots (`.`) at the end of a comment line, unless it is a sentence. Do start comments with a capital, though.

File: lib/ramfs/ramfs.h [hunk 8]
```diff
@@ -36,29 +36,65 @@
...
 	int rn_mode;
+	/* Whether or not the rn_buf was allocated in this ramfs_node. */
 	bool rn_owns_buf;
 };
 
+/**
+ * Allocates ramfs node.
+ *
+ * @param name
+ *   The name of the file/directory.
```
Comment (razvand) [inline lib/ramfs/ramfs.h]:
```suggestion
 *   The name of the file/directory
```

File: lib/ramfs/ramfs.h [hunk 9]
```diff
@@ -36,29 +36,65 @@
...
+ * @param type
+ *   The type (file or directory).
```
Comment (razvand) [inline lib/ramfs/ramfs.h]:
```suggestion
 *   The entry type (file or directory)
```

Comment (razvand) [reply lib/ramfs/ramfs.h]: What are possible values for `type`?

File: lib/ramfs/ramfs.h [hunk 10]
```diff
@@ -36,29 +36,65 @@
...
+ * @return
+ *   The new ramfs_node structure.
```
Comment (razvand) [inline lib/ramfs/ramfs.h]:
```suggestion
 *   Pointer to the new ramfs node
```

File: lib/ramfs/ramfs.h [hunk 11]
```diff
@@ -36,29 +36,65 @@
...
+ */
 struct ramfs_node *ramfs_allocate_node(const char *name, int type);
 
+/**
+ * Frees a ramfs node.
+ *
+ * @param node
+ *   The ramfs_node that will be freed.
```
Comment (razvand) [inline lib/ramfs/ramfs.h]:
```suggestion
 *   Pointer to the ramfs node that will be freed
```

File: lib/ramfs/ramfs.h [hunk 12]
```diff
@@ -36,29 +36,65 @@
...
+ */
 void ramfs_free_node(struct ramfs_node *node);
 
+/**
+ * Transforms a vnode into a ramfs_node.
+ *
+ * @param vnode
+ *   The virtual node (abstraction offered by vfscore)
+ *   which holds a reference to a ramfs_node.
```
Comment (razvand) [inline lib/ramfs/ramfs.h]:
```suggestion
 *   Virtual node (abstraction offered by vfscore) holding a reference to a ramfs_node
```

File: lib/ramfs/ramfs.h [hunk 13]
```diff
@@ -36,29 +36,65 @@
 #include <vfscore/prex.h>
 #include <stdbool.h>
 
-/*
- * File/directory node for RAMFS
+/**
+ * struct ramfs_node - A file/directory node for RamFS.
```
Comment (razvand) [inline lib/ramfs/ramfs.h]: We should be consistent in using `RamFS` or `RAMFS` or `ramfs` throughout the file.

File: lib/ramfs/ramfs.h [hunk 14]
```diff
@@ -36,29 +36,65 @@
...
 struct ramfs_node {
-	struct ramfs_node *rn_next;   /* next node in the same directory */
-	struct ramfs_node *rn_child;  /* first child node */
-	int rn_type;    /* file or directory */
-	char *rn_name;    /* name (null-terminated) */
-	size_t rn_namelen;    /* length of name not including terminator */
-	size_t rn_size;    /* file size */
-	char *rn_buf;    /* buffer to the file data */
-	size_t rn_bufsize;    /* allocated buffer size */
+	/* The next node in the same directory. */
```
Comment (razvand) [inline lib/ramfs/ramfs.h]: I would remove the `The` in the comments.

**Review by razvand — CHANGES_REQUESTED** [review]
Hi, @DeliaPavel . See my comments.

File: lib/ramfs/ramfs.h (hunk 13 above)
Comment (DeliaPavel) [reply lib/ramfs/ramfs.h]: How do you recommend we change this? `ramfs` is used when referring to the name of the structure `ramfs_node`, `RAMFS` is used for the macro `RAMFS_NODE`, and in `ifndef`, `define`, and `endif`. Should we change `RamFS` to one of these?

File: lib/ramfs/ramfs.h [hunk 15]
```diff
@@ -36,29 +36,66 @@
...
+	/* Last modification time */
 	struct timespec rn_mtime;
+	/* Node access mode */
 	int rn_mode;
+	/* Whether or not the rn_buf was allocated in this ramfs_node */
 	bool rn_owns_buf;
 };
 
+/**
+ * Allocates ramfs node.
```
Comment (razvand) [inline lib/ramfs/ramfs.h]:
```suggestion
 * Allocates a new ramfs node.
```

File: lib/ramfs/ramfs.h [hunk 16]
```diff
@@ -36,29 +36,66 @@
...
 void ramfs_free_node(struct ramfs_node *node);
 
+/**
+ * Transforms a vnode into a ramfs_node.
+ *
+ * @param vnode
+ *   Virtual node (abstraction offered by vfscore)
+ *   holding a reference to a ramfs_node
+ */
 #define RAMFS_NODE(vnode) ((struct ramfs_node *) vnode->v_data)
```
Comment (razvand) [inline lib/ramfs/ramfs.h]:
```suggestion
#define RAMFS_NODE(vnode) ((struct ramfs_node *) (vnode->v_data))
```

File: lib/ramfs/ramfs.h [hunk 17]
```diff
@@ -36,29 +36,66 @@
...
 	int rn_mode;
+	/* Whether or not the rn_buf was allocated in this ramfs_node */
 	bool rn_owns_buf;
 };
 
+/**
+ * Allocates ramfs node.
+ *
+ * @param name
+ *   The name of the file/directory
```
Comment (razvand) [inline lib/ramfs/ramfs.h]:
```suggestion
 *   The entry name
```

File: lib/ramfs/ramfs.h [hunk 18]
```diff
@@ -36,29 +36,66 @@
...
+ * @param type
+ *   The entry type (regular file - VREG, a symbolic link - VLNK,
+ *   or a directory - VDIR)
```
Comment (razvand) [inline lib/ramfs/ramfs.h]:
```suggestion
 *   The entry type (regular file - VREG, symbolic link - VLNK,
 *   or directory - VDIR)
```

File: lib/ramfs/ramfs.h [hunk 19]
```diff
@@ -36,29 +36,66 @@
...
-	size_t rn_size;    /* file size */
-	char *rn_buf;    /* buffer to the file data */
-	size_t rn_bufsize;    /* allocated buffer size */
+	/* Next node in the same directory */
+	struct ramfs_node *rn_next;
+	/* First child node */
+	struct ramfs_node *rn_child;
+	/* Type: file or directory */
+	int rn_type;
+	/* Name of the file/directory (null-terminated) */
```
Comment (razvand) [inline lib/ramfs/ramfs.h]:
```suggestion
	/* Name of the file/directory (NUL-terminated) */
```

File: lib/ramfs/ramfs.h [hunk 20]
```diff
@@ -36,29 +36,66 @@
...
-	char *rn_name;    /* name (null-terminated) */
-	size_t rn_namelen;    /* length of name not including terminator */
-	size_t rn_size;    /* file size */
-	char *rn_buf;    /* buffer to the file data */
-	size_t rn_bufsize;    /* allocated buffer size */
+	/* Next node in the same directory */
+	struct ramfs_node *rn_next;
+	/* First child node */
+	struct ramfs_node *rn_child;
+	/* Type: file or directory */
```
Comment (razvand) [inline lib/ramfs/ramfs.h]: I would add here the possible values: `VREG`, `VLNK`, `VDIR` and their explanation.

File: lib/ramfs/ramfs.h [hunk 21]
```diff
@@ -36,29 +36,66 @@
...
+ *   Pointer to the ramfs_node that will be freed
+ */
 void ramfs_free_node(struct ramfs_node *node);
 
+/**
+ * Transforms a vnode into a ramfs_node.
+ *
+ * @param vnode
+ *   Virtual node (abstraction offered by vfscore)
+ *   holding a reference to a ramfs_node
```
Comment (razvand) [inline lib/ramfs/ramfs.h:103]:
```suggestion
 *   holding a reference to a ramfs_node
 
 * @return
 *   Pointer to the corresponding ramfs_node
```

File: lib/ramfs/ramfs.h [hunk 22]
```diff
@@ -36,29 +36,66 @@
 #include <vfscore/prex.h>
 #include <stdbool.h>
 
-/*
- * File/directory node for RAMFS
+/**
+ * struct ramfs_node - A file/directory node for RamFS
```
Comment (razvand) [inline lib/ramfs/ramfs.h]:
```suggestion
 * struct ramfs_node - A filesystem entry node for RamFS
```

**Review by razvand — CHANGES_REQUESTED** [review]
@DeliaPavel , new review round

**Review by StefanJum — APPROVED** [review]
Looks all good to me @DeliaPavel, thank you, I'll wait for @razvand to take another look and then add the tag.

File: lib/ramfs/ramfs.h [hunk 23]
```diff
@@ -36,29 +36,72 @@
...
-	struct ramfs_node *rn_child;  /* first child node */
-	int rn_type;    /* file or directory */
-	char *rn_name;    /* name (null-terminated) */
-	size_t rn_namelen;    /* length of name not including terminator */
-	size_t rn_size;    /* file size */
-	char *rn_buf;    /* buffer to the file data */
-	size_t rn_bufsize;    /* allocated buffer size */
+	/* Next node in the same directory */
+	struct ramfs_node *rn_next;
+	/* First child node */
```
Comment (razvand) [inline lib/ramfs/ramfs.h]: What if the current node is not a directory? Would this be `NULL`? If so, this should be part of the comment.

**Review by razvand — CHANGES_REQUESTED** [review]
@DeliaPavel, see my minor comment. Then we can merge this.

**Review by razvand — APPROVED** [review]
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Review by StefanJum — APPROVED** [review]
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

**Pull request closed without merging**